spinnaker-monitoring-daemon/spinnaker-monitoring/meter_specification_handler.py

```diff
@@ -0,0 +1,498 @@
+        result.append('  "%s"' % meter)
+        result.append('    * %s' % ('\n    * '.join(remarks)))
+        result.append('')  # for separator between meter entries
+      return '\n'.join(result)
+
+    text = '\n'.join([
+        service_remarks_to_string(service, meter_diff)
+      for service, meter_diff in service_diff.items()
+        if meter_diff
+    ])
+    self.output(options, text)
+
+
+class AugmentMeterSpecificationsHandler(BaseSpectatorCommandHandler):
+  """Various heuristics to augment meter specs with more info."""
+
+  def add_argparser(self, subparsers):
+    abspath = os.path.abspath(__file__)
+    prometheus_dir = os.path.join(
+        os.path.dirname(os.path.dirname(os.path.dirname(abspath))),
+        'spinnaker-monitoring-third-party',
+        'third_party',
+        'prometheus')
+
+    parser = super(AugmentMeterSpecificationsHandler, self).add_argparser(
+        subparsers)
+    parser.add_argument(
+        '--baseline', default=None,
+        help='If specified, is either a single yaml file keyed by service,'
+             ' or a directory with yaml files named'
+             ' <service>-meter-specification.yml')
+    parser.add_argument(
+        '--input', default=None, required=True,
+        help='Provides then meter specification attributes to augment.'
+             ' Either a single yaml file keyed by service,'
+             ' or a directory with yaml files named'
+             ' <service>-meter-specification.yml')
+    parser.add_argument(
+        '--per_service_output_dir', default=None,
+        help='Write a yaml file for each service'
+             ' named <service>-meter-specification.yml')
+    parser.add_argument(
+        '--prometheus_dashboard_dir', default=prometheus_dir,
+        help='If specified infer viewpoints from prometheus dashboards.')
+    return parser
+
+  def process_commandline_request(self, options):
+    all_baseline = load_all_service_meter_specs(
+        options.get('baseline'))
+    all_changes = load_all_service_meter_specs(
+        options.get('input'))
+
+    viewpoint_augmenter = ViewpointAugmenter(options)
+    yaml_dict = {}
+    for service in union_keys(all_baseline, all_changes):
+      logging.info('Processing service="%s"', service)
+      baseline = all_baseline.get(service)
+      changes = all_changes.get(service)
+      yaml_dict[service] = self.fuse_all(baseline, changes)
+      viewpoint_augmenter.augment(service, yaml_dict[service])
+
+    dump_all_meter_spec(options.get('per_service_output_dir'), yaml_dict)
+    self.output(options, encode_yaml(yaml_dict))
+
+  def fuse(self, baseline, current):
+    """Fuse the meter spec from current into baseline."""
+    spec = dict(baseline)
+    for key, value in current.items():
+      if not key in baseline:
+        spec[key] = value
+        continue
+      if isinstance(value, dict):
+        spec[key].update(value)
+        continue
+      if isinstance(value, list):
```

Comment:
I'm guessing these lists won't be too long, but this is O(N^2) in the size of the lists

Reply:
Lists are not that long. Simplicity is more important than runtime efficiency here. This is offline.